examples/lessons/13/app.js

```diff
@@ -0,0 +1,294 @@
+attribute vec3 normals;
+attribute vec2 texCoords;
+
+uniform mat4 uMMatrix;
+uniform mat4 uVMatrix;
+uniform mat4 uPMatrix;
+
+uniform vec3 uAmbientColor;
+
+uniform vec3 uPointLightingLocation;
+uniform vec3 uPointLightingColor;
+
+uniform bool uUseLighting;
+
+varying vec2 vTextureCoord;
+varying vec3 vLightWeighting;
+
+void main(void) {
+  vec4 mPosition = uMMatrix * vec4(positions, 1.0);
+  gl_Position = uPMatrix * uVMatrix * mPosition;
+  vTextureCoord = texCoords;
+
+  if (!uUseLighting) {
+    vLightWeighting = vec3(1.0, 1.0, 1.0);
+  } else {
+    vec3 lightDirection = normalize(uPointLightingLocation - mPosition.xyz);
+    vec4 transformedNormal = uMMatrix * vec4(normals, 0.0);
+    float pointLightWeighting = max(dot(transformedNormal.xyz, lightDirection), 0.0);
+    vLightWeighting = uAmbientColor + uPointLightingColor * pointLightWeighting;
+  }
+}
+`;
+
+const VERTEX_LIGHTING_FRAGMENT_SHADER = `\
+#ifdef GL_ES
+precision highp float;
+#endif
+
+varying vec2 vTextureCoord;
+varying vec3 vLightWeighting;
+
+uniform bool uUseTextures;
+uniform sampler2D uSampler;
+
+void main(void) {
+  vec4 fragmentColor;
+  if (uUseTextures) {
+    fragmentColor = texture2D(uSampler, vec2(1.0 - vTextureCoord.s, 1.0 - vTextureCoord.t));
+  } else {
+    fragmentColor = vec4(1.0, 1.0, 1.0, 1.0);
+  }
+  gl_FragColor = vec4(fragmentColor.rgb * vLightWeighting, fragmentColor.a);
+}
+`;
+
+const FRAGMENT_LIGHTING_VERTEX_SHADER = `\
+precision highp float;
+attribute vec3 positions;
+attribute vec3 normals;
+attribute vec2 texCoords;
+
+uniform mat4 uMMatrix;
+uniform mat4 uVMatrix;
+uniform mat4 uPMatrix;
+
+varying vec2 vTextureCoord;
+varying vec4 vTransformedNormal;
+varying vec4 vPosition;
+
+void main(void) {
+  vPosition = uMMatrix * vec4(positions, 1.0);
+  gl_Position = uPMatrix * uVMatrix * vPosition;
+  vTextureCoord = texCoords;
+  vTransformedNormal = uMMatrix * vec4(normals, 0.0);
+}
+`;
+
+const FRAGMENT_LIGHTING_FRAGMENT_SHADER = `\
+#ifdef GL_ES
+precision highp float;
+#endif
+
+uniform vec3 uAmbientColor;
+
+uniform vec3 uPointLightingLocation;
+uniform vec3 uPointLightingColor;
+
+uniform bool uUseLighting;
+uniform bool uUseTextures;
+
+varying vec2 vTextureCoord;
+varying vec4 vTransformedNormal;
+varying vec4 vPosition;
+
+uniform sampler2D uSampler;
+
+void main(void) {
+  vec3 lightWeighting;
+
+  if (!uUseLighting) {
+    lightWeighting = vec3(1.0, 1.0, 1.0);
+  } else {
+    vec3 lightDirection = normalize(uPointLightingLocation - vPosition.xyz);
+    float pointLightWeighting = max(dot(vTransformedNormal.xyz, lightDirection), 0.0);
+    lightWeighting = uAmbientColor + uPointLightingColor * pointLightWeighting;
+  }
+
+  vec4 fragmentColor;
+  if (uUseTextures) {
+    fragmentColor = texture2D(uSampler, vec2(1.0 - vTextureCoord.s, 1.0 - vTextureCoord.t));
+  } else {
+    fragmentColor = vec4(1.0, 1.0, 1.0, 1.0);
+  }
+  gl_FragColor = vec4(fragmentColor.rgb * lightWeighting, fragmentColor.a);
+}
+`;
+
+const moonRotation = {
```

Comment:
Why wrap this in an object? Can't we just do
```
const moonRotation = new Matrix4()....
```

Reply:
it doesn't work in Safari by using "const moonRotation = new Matrix4()....", cause global functions can't access const or let in Safari even I pass const or let as parameters to them. After wrapping it with an object, it works well.

Reply:
> After wrapping it with an object, it works well.

Got it. In that case I would suggest using a single wrapper 
```
const appState = {
  moonRotation: new ...,
  cubeRotation: 
};
```
It would serve the purpose of holding all app state together rather than just being a confusing work-around for a browser issue.

Maybe update all examples with this approach?

Reply:
No problem. I will do it.